### ISSUE TITLE
Random NullReferenceException on object.GetType

### ISSUE THREAD

**Issue by trivalik** [nhibernate/nhibernate-core#3342]
I have a random (one time right now) failure:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Object.GetType()
   at xxx.yyy.zzz.SaveIt.SaveSomeData(Meta meta, IEnumerable`1 metaData, ConcurrentBag`1 additonalInfo)
   at System.Threading.Tasks.Task.Execute()
```

In `SaveSomeData` I just save some objects with NHibernate 3.3.1. Therefore OpenStatelessSession and also OpenSession is used. No property access on the objects which are provide to this method. I can just imagine that this happens through the ReflectionOptimizer and the CallVirt emit, here https://github.com/nhibernate/nhibernate-core/blob/733032976f07c61b4cffd98c71f7b37c0af025f4/src/NHibernate/Bytecode/Lightweight/ReflectionOptimizer.cs#L169C22-L169C30

Can I null check in emit helps here?

Why shows my callstack no NHibernate? maybe it was because it was built as release?

I posted this because newer versions has same callvirt.

edit: I ran in similar error, which is probably the same. Here a more detailed stacktrace:

```
System.AggregateException: One or more errors occurred. An error occurred while sending the request. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Object.GetType()
   at NHibernate.Connection.ConnectionProvider.CloseConnection(IDbConnection conn)
   at NHibernate.Connection.DriverConnectionProvider.CloseConnection(IDbConnection conn)
   at NHibernate.Transaction.AdoNetTransactionFactory.ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, Boolean transacted)
   at NHibernate.Transaction.AdoNetWithDistributedTransactionFactory.ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, Boolean transacted)
   at NHibernate.Engine.TransactionHelper.DoWorkInNewTransaction(ISessionImplementor session)
   at NHibernate.Id.TableGenerator.Generate(ISessionImplementor session, Object obj)
   at NHibernate.Id.TableHiLoGenerator.Generate(ISessionImplementor session, Object obj)
   at NHibernate.Impl.StatelessSessionImpl.Insert(String entityName, Object entity)
   at NHibernate.Impl.StatelessSessionImpl.Insert(Object entity)
   at xxx.yyy.zzz.SaveIt.SaveSomeData(Meta meta, IEnumerable`1 metaData, ConcurrentBag`1 additonalInfo) in C:\xxx\SaveIt.cs:Line209.
   at xxx.yyy.zzz.SaveIt.<>c__DisplayClass12_0.<RunInternal>b__1() in C:\xxx\SaveIt.cs:Line 160.
   bei System.Threading.Tasks.Task.Execute()
```

**Comment by trivalik** [comment]
After adding the details stacktrace, I see that the null check is missing. Could you fix it on your own? I am not sure if now no ADOException should be thrown or not.

**Comment by bahusoid** [comment]
There is already a null check throwing `ArgumentNullException`.
https://github.com/nhibernate/nhibernate-core/blob/e2e6cbe78325d8495931b18b64c77bfc520eaf29/src/NHibernate/Connection/ConnectionProvider.cs#L27-L28

Maybe it's not present in 3.3 but no one is going to investigate obsolete version. Please let us know when you have some issues on recent NHibernate versions.

**Comment by trivalik** [comment]
Sorry I missed that check.

**Comment by fredericDelaporte** [comment]
For reference, the current version is 5.4.x. Fixes for things which are not a regression compared to a 5.x release are done in the next future version not yet released, 5.5.0.
We are currently still actively maintaining 5.3.x for newly spotted regressions compared to a previous 5.x version. We may do newer 5.0.x, 5.1.x or 5.2.x releases for critical (but very unlikely to be not yet spotted) regressions or security issues. We do not consider investigating troubles for older releases.
